jetty-ee10/jetty-ee10-webapp/src/test/java/org/eclipse/jetty/ee10/webapp/WebAppContextTest.java

```diff
@@ -791,6 +800,99 @@ public void testExtraClasspathDir(String extraClassPathReference) throws Excepti
         assertThat("URL[0]", urls[0].toURI(), is(extLibs.toUri()));
     }
 
+    @Test
+    public void testRestartWebApp(WorkDir workDir) throws Exception
+    {
+        Server server = newServer();
+
+        // Create war
+        Path tempDir = workDir.getEmptyPathDir();
+        Path testWebappDir = MavenPaths.projectBase().resolve("src/test/webapp");
+        assertTrue(Files.exists(testWebappDir));
+        Path warFile = tempDir.resolve("demo.war");
+
+        Map<String, String> env = new HashMap<>();
+        env.put("create", "true");
+
+        URI uri = URI.create("jar:" + warFile.toUri().toASCIIString());
+        // Use ZipFS so that we can create paths that are just "/"
+        try (FileSystem zipfs = FileSystems.newFileSystem(uri, env))
+        {
+            Path root = zipfs.getPath("/");
+            IO.copyDir(testWebappDir, root);
+        }
+
+        // Create WebAppContext
+        WebAppContext context = new WebAppContext();
+        ResourceFactory resourceFactory = context.getResourceFactory();
+        Resource warResource = resourceFactory.newResource(warFile);
+        context.setContextPath("/");
+        context.setWarResource(warResource);
+        context.setExtractWAR(true);
```

Comment:
I'm not sure what this test is testing? As you don't set the temporary directory, the webapp will be extracted into a freshly made tmp dir each time and the problem from #11549 will never occur anyway.

Reply:
OK so these are different problems altogether. 

The first problem we identified is the problem with the jvm internals' lack of ability to clean up its file system references if the jar file no longer exists. So we need one one fix for that, which is a combination of @lorban  changes in #11549, and a one line change to move the point at which the webapp temporary directory is deleted. Either @lorban adds this one line to his PR, or you just do it in this PR, but without other changes.

Then, raise a new PR with this test, which is doing something else entirely - it is stopping and restarting _an already existing context_.  That may well reveal other problems that need fixing, but we need to clearly identify what those problems are and how to fix them. From what you're saying it seems to me that the context isn't being cleaned up sufficiently on a stop so I don't think the solution is to check for duplicate jars etc etc.  We've always been able to do a context stop/restart on previous versions of jetty without leaking memory or classloaders, so we need to compare the new behaviour in 12 with what we have done previously to understand where any problems might lay.

Reply:
FWIW, I copied your new `WebAppContextTest.testRestartWebApp` test into jetty-12.0.x head, moved the `stopWebApp()` line in `WebAppContext` as you have above and the test worked without errors, no changes to `WebAppClassLoader` nor `MetaInfConfiguration` required.

Interestingly, before I moved the `stopWebApp()` line, I did get the `FileSystemClosedException`, which I wasn't really expecting to see because jetty is picking the temporary directory. Maybe when we stop the context, we aren't clearing out the name of the temporary directory? If jetty is picking it, I think we should clear it. I'll check what we did in jetty-11 etc.

Reply:
See #11558

Reply:
@joakime we still desperately need a PR that simply changes the position of the call to `stopWebApp()` - can you either change this into that, or make a new PR and submit?

Reply:
OK.
This PR has minimal changes now.
Lets see if the CI build goes green.